### ISSUE TITLE
parser issues related to RESTART and SKIPREST

### ISSUE THREAD

**Issue by totto82** [OPM/opm-parser#773]
The parser throws an error if SKIPREST is present in the deck without RESTART. Eclipse ignores SKIPREST if RESTART is not in the deck.

**Comment by andlaus** [comment]
are the SKIPREST, SKIP100 and SKIP300 keywords supported by opm-parser at all?

**Comment by joakim-hove** [comment]
Comment copied from:https://github.com/OPM/opm-core/issues/1004

Currently FLOW computes the initial state also for restart runs. This would not be a big problem ( other than some wasted calculations), if not for the fact that somehow Eclipse gives an error if EQUIL or other initialization related keywords is specified in Restart runs.

**Comment by joakim-hove** [comment]
> are the SKIPREST, SKIP100 and SKIP300 keywords supported by opm-parser at all?

The `SKIP100` and `SKIP300` keywords are not supported at all; it is on the loose-way-into-future plan to implement something like `ParseIdentity` so the parser can know: "Allright - now I am parsing as e.g. ECLIPSE100 and should behave accordingly" - then we can support `SKIP100` and `SKIP300`.

`SKIPREST` is a different beast entirely; I guess it's origins comes from the fact that (I assume) ECLIPSE really works works like an interpreter stepping through the input file one keyword at a time. When RESTART is active it should not "execute" the `TSTEP` or `DATES` keywords found in the Schedule section but rather skip forward until the restart point.

**Comment by joakim-hove** [comment]
OK;

I guess this will be a tradeoff between strict ECLIPSE behavior and sanity. As I see it there are typically three different paths:

1. What the Eclipse documentation says.
2. What Eclipse actually does.
3. What would be the sane thing to do.

As I see it the main problem here is that the Eclipse input format is used to both configure the case as a whole - and one particular run using the `RESTART / SKIPREST` keywords. Given a fully specified input deck it should be possible to instruct the simulator as:

```bash
bash% simulator DECK.DATA --restart=/path/to/restart_file:70
```

*Without* actually modifying the input deck. We can easily (and should?!) support this behavior, but that will be an extension of Eclipse.

**Eclipse restart semantics:** To instruct Eclipse to restart a simulation you must go into the deck and make the following modifications:

1. In the `SOLUTION` section you must insert the `RESTART` keyword:
```txt
RESTART
    /path/to/case/INITCASE 20 /
```

2. In the `SOLUTION` section you must remove the keywords used for initialization, be that `EQUIL` or something else. The point is that from an Eclipse point of view *the initialization* is replaced with *load from restart file*.

3. In the `SCHEDULE` section you must add the `SKIPREST` keyword.

SInce we properly internalize the complete input deck before actually starting the simulation we do not need any of these hoops. My suggestion is:

1. We respect the `RESTART` keyword - i.e. if that is found in the deck we do a restart.
2. If we are doing a restart we do not actually perform any initialization - but we do not require that the initialization keywords are removed from the input deck.
3. `SKIPREST` - we completely ignore this keyword. If the `RESTART`keyword is found in the deck this will be more lenient than Eclipse which requires `SKIPREST`to be present if `RESTART` is present - and it is in accordance with actual Eclipse behavior iff a lonesome `SKIPREST`is found.

Finally I also suggest that we support something like:
```bash
bash% simulator DECK.DATA --restart=/path/to/restart_file:70
```
which would be *deck independent* restart.

**Comment by totto82** [comment]
Sounds like a good plan. I agree that we should not do everything exactly the same way as eclipse, but the possibility to run the same deck in OPM and Eclipse without modification is important. Your suggestion gives us exactly this possibility without adding unnecessary noise.

**Comment by OPMUSER** [comment]
Joakim,

No the SKIPREST keyword is optional, you do have to use it.

Before the SKIPREST keyword was implemented one had to delete all (not 
all but most) of the schedule data up to and including the restart 
point. Using the SKIPREST keyword allows the engineer just to enter the 
SKIPREST in the input deck and ECLIPSE automatically ignores those time 
dependent keywords but includes the non-time dependent data like the VFP 
tables etc.

This has the advantage that you keep the input deck "whole" in that the 
only changes are to the SOLUTION section where the equilibrium data is 
replace by the RESTART keyword, plus the SKIPREST keyword normally 
placed at the beginning of the SCHEDULE section.

OPM should behave is the same way, that is treat the SKIPREST keyword as 
optional

 2. If we are doing a restart we do not actually perform any
    initialization - but we do not require that the initialization
    keywords are removed from the input deck.

    *This is the same behavior as ECLIPSE and should cause an error if
    EQUIL and RESTART are in the same deck. Note*

 3. SKIPREST - we completely ignore this keyword. If the RESTART keyword
    is found in the deck this will be more lenient than Eclipse which
    requires SKIPREST to be present if RESTART is present - and it is in
    accordance with actual Eclipse behavior iff a lonesome SKIPRESTis found.

    *No you cannot ignore the keyword, if it is present in a restart run
    then it should be processed. If it is absent then continue as
    normal, as the engineer may have made the required changes manually.
    *

It does not make sense have a command like restart facility, as you 
still need to modifying the input by deleting the equilibrium data, as 
per (2) above. If you allow a command line restart facility and the 
engineer does not amend the equilibrium data then you should throw an 
error. So now you have amend the deck to get it running anyway, so I 
only see confusion with this additional facility.
ThunderBird Signature File

Regards, David Baxendale Email: David.Baxendale@gmail.com

------------------------------------------------------------------------

**Comment by joakim-hove** [comment]
@OPMUSER : Thank you for contributing to the discussion. 

My take on the situation is:

1. If ECLIPSE will restart from a Deck we should also restart from the same deck.
2. If ECLIPSE reports an error in a restart deck due to either missing `SKRIPREST` or initialization keywords like `EQUIL` - we should not report an error, but rather restart according to the `RESTART`keyword.
3. We should support a restart capability which requires *zero* changes to deck, i.e. `flow --restart DECK.DATA`

My mental picture of Eclipse execution is an imperative programming language where the keywords in the input file are executed one by one; that is quite different from the model employed in opm/flow where the full input data is internalized before we actually start simulating. The difference in execution model allows us to make improvements as those mentioned in points 2) and 3).

**Comment by OPMUSER** [comment]
Joakim,

I understand what you wish to do but I have the following comments.

There is more than one way to run a restart job from ECLIPSE, for 
example using:

 1. the SAVE file option for the initial data and time dependent data,
 2. the SAVE file option for the initial data and the RESTART file for
    the time dependent data, and
 3. the input deck for the initial data and the RESTART file for the
    time dependent  time.

Now I know that Flow does not have the SAVE file option, but engineers 
will take their existing decks and run then in Flow. So if you allow for 
your option (2) and (3) you are going to have to check that all the 
other possibilities are flagged as errors. And remember at the moment 
there is no documentation on what keywords are support by Flow.

Secondly, ECLIPSE allows the engineers to stipulate what gets written to 
the RESTART file. Some people use this feature to save only the data 
required for graphics, so it is not a "proper" RESTART file. Does Flow 
check that the RESTART file is valid for restarting? Also the SKIPREST 
keyword is optional, if it is missing then there is no warning or error 
message.

Basically, restart runs in ECLIPSE are very flexible (or complicated to 
the inexperience user) so adding another option is jgoing to cause some 
issues.
ThunderBird Signature File

Regards, David Baxendale Email: David.Baxendale@gmail.com

------------------------------------------------------------------------

**Comment by joakim-hove** [comment]
> Currently FLOW computes the initial state also for restart runs. This would not be a big problem ( other than some wasted calculations), if not for the fact that somehow Eclipse gives an error if EQUIL or other initialization related keywords is specified in Restart runs.

@totto82: I am looking at this now; do not really understand. If `EQUIL` is removed from the deck - as a convenience for ECLIPSE compatibility - FLOW should still work fine?

**Comment by totto82** [comment]
No it doesn't. InitState_impl.hpp:735 in opm-core complains. As it wants either EQUIL or PRESSURE, SWAT/SOIL/SGAS in the deck.

**Comment by atgeirr** [comment]
Refactoring the initialization code is something I'd wanted to do for a while, but have not had the time. There are probably some major simplifications that can be made there. However this particular problem might be solved on a higher level, not entering that code part at all if we initialize from a restart file.

**Comment by joakim-hove** [comment]
> Refactoring the initialization code is something I'd wanted to do for a while

Yes please do - it is quite difficult to navigate. Case in point: I used quite a lot of time finding the call path from `FlowMain.hpp` to the offending call site: `InitState_impl.hpp:735`. The current initialization code consists of a mix of nontriivial code to determine *which* initialization approach to use, and the actual code to perform the initialization - it is a mess. I think readability would improve considerably if these two were separated 100%:

1. Inspect your input and determine which algorithm to use.
2. Extract the necessary input[1] and call the right algorithm.

For the current code `FlowMain.hpp:507` will check if the deck has keyword `EQUIL`- failing that it will enter the block on lines 516-527; which will again check if the deck has `EQUIL`- and fail hard on the location mentioned by @totto82. 

> However this particular problem might be solved on a higher level, not entering that code part at all if we initialize from a restart file.

Well - that was certainly my intention, but the initialization code seems so convoluted that I think the chance of getting some other initialization wrong is quite high. This is a typical case where an  Insufficient test suite makes developers who  not very familiar with the code afraid to make changes - and progress stops.

[1]: Passing `Deck&` instances around will soon become illegal - the initialization code is currently full of it. The `EclipseState`has classes `Equil` and `ThresholdPressure` created specifically to simplify the initialization code - seems we have not utilized these properly yet.

**Comment by OPMUSER** [comment]
Note that ECLIPSE allows the engineer to modify the initialization data 
when running a flexible restart run (initialization data in the deck and 
a RESTART file). Don't known why anybody would want to change the 
property data this way, but somebody must have requested this feature. 
One can also change some of the dimensional data in the RUNSPEC section 
by decreasing the footprint of the run (cannot increase the parameters)
ThunderBird Signature File

Regards, David Baxendale Email: David.Baxendale@gmail.com

------------------------------------------------------------------------

**Comment by jokva** [comment]
https://github.com/OPM/opm-parser/pull/963
